test/misc.jl

```diff
@@ -158,6 +158,32 @@ for l in (Threads.SpinLock(), ReentrantLock())
     @test try unlock(l) finally end === nothing
 end
 
+@testset "Semaphore" begin
+    sleep_secs = 0.01
+    sem_size = 2
+    n = 100
+
+    s = Base.Semaphore(sem_size)
+
+    t = @elapsed @sync for _ in 1:n
+        @async begin
+            Base.acquire(s)
+            sleep(sleep_secs)
+            Base.release(s)
+        end
+    end
+    @test isapprox(t, (n / sem_size) * sleep_secs, rtol = 0.2)
+
+    t = @elapsed @sync for _ in 1:n
+        @async begin
+            Base.acquire(s) do
+                sleep(sleep_secs)
+            end
+        end
+    end
+    @test isapprox(t, (n / sem_size) * sleep_secs, rtol = 0.2)
```

Comment:
Testing this based on run-time is a bit fragile.  How about checking the invariance with

```julia
clock = Threads.Atomic{Int}(1)
occupied = Threads.Atomic{Int}(0)
history = zeros(Int, 2n)
```

and then something like

```julia
Base.acquire(s) do
    history[Threads.atomic_add!(clock, 1)] = Threads.atomic_add!(occupied, 1)
    sleep(rand(0:0.01:0.1))
    history[Threads.atomic_add!(clock, 1)] = Threads.atomic_sub!(occupied, 1)
end
```

in each task. Then, we expect

```julia
@test all(<=(sem_size), history)
```

Reply:
Thanks! Yeah, I thought something like this would be better but hadn't figured it out. Pushed